8261137: Optimization of Box nodes in uncommon_trap

Wanghuang-Huawei opened the pull request:
JDK-8075052 has  removed useless autobox. However, in some cases, the box is still saved. For instance:
```
@Benchmark
public void testMethod(Blackhole bh) {
  int sum = 0;
  for (int i = 0; i < data.length; i++) {
      Integer ii = Integer.valueOf(data[i]);
      if (i < data.length) {
          sum += ii.intValue();
      }
  }
  bh.consume(sum);
}
```
Although the variable ii is only used at ii.intValue(), it cannot be eliminated as a result of being used by a hidden uncommon_trap.
The uncommon_trap is generated by the optimized "if", because its condition is always true.

We can postpone box in uncommon_trap in this situation. We treat box as a scalarized object by adding a SafePointScalarObjectNode in the uncommon_trap node,
and deleting the use of box:

There is no additional fail/error(s) of jtreg after this patch.

I adjust my codes and add a new benchmark

```java
public class MyBenchmark {

    static int[] data = new int[10000];

    static {
        for(int i = 0; i < data.length; ++i) {
            data[i] = i * 1337 % 7331;
        }
    }

    @Benchmark
    public void testMethod(Blackhole bh) {
      int sum = 0;
      for (int i = 0; i < data.length; i++) {
          Integer ii = Integer.valueOf(data[i]);
          black();
          if (i < 100000) {
              sum += ii.intValue();
          }
      }
      bh.consume(sum);
    }

    public void black(){}
}
```


aarch64:  
base line：
```
Benchmark                     Mode  Samples   Score  Score error  Units
o.s.MyBenchmark.testMethod    avgt       30  88.513        1.111  us/op
```

opt：
```
Benchmark                     Mode  Samples   Score  Score error  Units
o.s.MyBenchmark.testMethod    avgt       30  52.776        0.096  us/op
```

x86:  
base line：
```
Benchmark                     Mode  Samples   Score  Score error  Units
o.s.MyBenchmark.testMethod    avgt       30  81.066        3.156  us/op
```

opt:
```
Benchmark                     Mode  Samples   Score  Score error  Units
o.s.MyBenchmark.testMethod    avgt       30  55.596        0.775  us/op
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261137](https://bugs.openjdk.java.net/browse/JDK-8261137): Optimization of Box nodes in uncommon_trap


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 3934fd9a924d28b524b7108299a3ab17bcfb414e
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)


### Contributors
 * Wu Yan `<wuyan34@huawei.com>`
 * Ai Jiaming `<aijiaming1@huawei.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2401/head:pull/2401` \
`$ git checkout pull/2401`

Update a local copy of the PR: \
`$ git checkout pull/2401` \
`$ git pull https://git.openjdk.java.net/jdk pull/2401/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2401`

View PR using the GUI difftool: \
`$ git pr show -t 2401`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2401.diff">https://git.openjdk.java.net/jdk/pull/2401.diff</a>

</details>

Wanghuang-Huawei commented:
/contributor add Wu Yan <wuyan34@huawei.com>

Wanghuang-Huawei commented:
/contributor add Ai Jiaming <aijiaming1@huawei.com>

Wanghuang-Huawei commented:
/contributor add Ai Jiaming <aijiaming1@huawei.com>

theRealELiu commented:
I was wandering if we can remove the useless `if` as it's always true in this case. Do you know why this kind of `if` haven't been eliminated by GVN phase?

neliasso commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -603,6 +603,51 @@ void CallGenerator::do_late_inline_helper() {
     C->remove_macro_node(call);
   }
 
+  if (callprojs.resproj != NULL && call->is_CallStaticJava() &&
```
Please extract the new code into a method with a descriptive name.

Wanghuang-Huawei replied:
OK. I will refactor these codes ASAP.

Wanghuang-Huawei commented:
> I was wandering if we can remove the useless `if` as it's always true in this case. Do you know why this kind of `if` haven't been eliminated by GVN phase?
It is a good idea . However, C2 might not optimize some `if` for many reasons: 
*  This situation can also be triggered in calling other method. For example :
  ```
public class MyBenchmark {

    static int[] data = new int[10000];

    static {
        for(int i = 0; i < data.length; ++i) {
            data[i] = 299;
        }
    }

    @Benchmark
    public void testMethod(Blackhole bh) {
      int sum = 0;
      for (int i = 0; i < data.length; i++) {
        Integer ii = Integer.valueOf(data[i]);
        sum += abs(ii);
      }
      bh.consume(sum);
    }

    public int abs(Integer ii) {
      if (ii.intValue() > 0) {
        return ii.intValue();
      } else {
        return 0 - ii.intValue();
      }
    }
}
```
 The method `abs` will be inlined when it is hot enough. However, this `If` can not be eliminated.
*  The case in the JDK-8261137 is just like this case: 
```
public class MyBenchmark {

    static int[] data = new int[10000];

    static {
        for(int i = 0; i < data.length; ++i) {
            data[i] = i * 1337 % 7331;
        }
    }

    @Benchmark
    public void testMethod(Blackhole bh) {
      int sum = 0;
      for (int i = 0; i < data.length; i++) {
          Integer ii = Integer.valueOf(data[i]);
          if (i < 100000) {
              sum += ii.intValue();
          }
      }
      bh.consume(sum);
    }
}
```

theRealELiu commented:
> > I was wandering if we can remove the useless `if` as it's always true in this case. Do you know why this kind of `if` haven't been eliminated by GVN phase?

> It is a good idea . However, C2 might not optimize some `if` for many reasons:
> 
> * This situation can also be triggered in calling other method. For example :
> 
> ```
> public class MyBenchmark {
> 
>   static int[] data = new int[10000];
> 
>   static {
>       for(int i = 0; i < data.length; ++i) {
>           data[i] = 299;
>       }
>   }
> 
>   @Benchmark
>   public void testMethod(Blackhole bh) {
>     int sum = 0;
>     for (int i = 0; i < data.length; i++) {
>       Integer ii = Integer.valueOf(data[i]);
>       sum += abs(ii);
>     }
>     bh.consume(sum);
>   }
> 
>   public int abs(Integer ii) {
>     if (ii.intValue() > 0) {
>       return ii.intValue();
>     } else {
>       return 0 - ii.intValue();
>     }
>   }
> }
> ```
> 
> The method `abs` will be inlined when it is hot enough. However, this `If` can not be eliminated.
> 
> * The case in the JDK-8261137 is just like this case:
> 
> ```
> public class MyBenchmark {
> 
>     static int[] data = new int[10000];
> 
>     static {
>         for(int i = 0; i < data.length; ++i) {
>             data[i] = i * 1337 % 7331;
>         }
>     }
> 
>     @Benchmark
>     public void testMethod(Blackhole bh) {
>       int sum = 0;
>       for (int i = 0; i < data.length; i++) {
>           Integer ii = Integer.valueOf(data[i]);
>           if (i < 100000) {
>               sum += ii.intValue();
>           }
>       }
>       bh.consume(sum);
>     }
> }
> ```

Thanks for your explanation. This makes more sense to me now.

vnkozlov commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,53 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+        ProjNode* res = resproj->as_Proj();
+        Node* uncommon_trap_node = delay_boxes.pop();
+        int in_edge = uncommon_trap_node->find_edge(res);
+        assert(in_edge > 0, "sanity");
```
If there are several references you need to replace all of them.

navyxliu replied:
+1
scalar replacement uses range substitution
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/macro.cpp#L892

Wanghuang-Huawei replied:
Thank you for your review. It's my fault. I will revise this in next commit.

vnkozlov commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,53 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+        uint first_ind = (uncommon_trap_node->req() - uncommon_trap_node->jvms()->scloff());
+        Node* sobj = new SafePointScalarObjectNode(gvn.type(res)->isa_oopptr(),
+#ifdef ASSERT
+                                                  NULL,
```
I would suggest to record `call` node here treating it as allocation.

vnkozlov reviewed (commented):
I am a little concern about stretching uses of input value outside scope where it is created (for example, loop's variable or a value depending on it).
This optimization may work only because boxed values are immutable.
I will run our tests with this changes.

vnkozlov commented:
Our tier1-4 testing passed

navyxliu commented:
> I was wandering if we can remove the useless `if` as it's always true in this case. Do you know why this kind of `if` haven't been eliminated by GVN phase?

Eventually, c2 should know that `i < data.length' is true. it should be done by GCP later. 

The example is a special case.  you can remove this " if (i < data.length)"  here.  Generally, target code can look like this. c2 speculatively generates an uncommon_trap of "unstable_if".

```
  for (int i = 0; i < data.length; i++) {
      Integer ii = Integer.valueOf(data[i]);
      if (cond) { //likely
          sum += ii.intValue();
      }
  }

```

navyxliu commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,53 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
 }
 
+static void delay_box_in_uncommon_trap(CallNode* call, Node* resproj) {
+  if (resproj != NULL && call->is_CallStaticJava() &&
```
IMHO, we should use nullptr here because hotspot now is using c++14.

Wanghuang-Huawei replied:
Thank you for your review. I will change that.

navyxliu commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,53 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+static void delay_box_in_uncommon_trap(CallNode* call, Node* resproj) {
+  if (resproj != NULL && call->is_CallStaticJava() &&
+      call->as_CallStaticJava()->is_boxing_method()) {
+    GraphKit kit(call->jvms());
```
you can postpone to construct this object in if(no_use).

Wanghuang-Huawei replied:
Sure. I will change that.

navyxliu commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,53 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+
+        ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
+        int n_fields = klass->nof_nonstatic_fields();
+        assert(n_fields == 1, "sanity");
```
I think you  also need to check the only non-static field of klass must be a scalar.
"sanity" is too concise. I think we should leave a message to say it's an auto-boxing class.

iwanowww reviewed (changes requested):
The improvement you are proposing is not specific to uncommon traps, but can be generalized to any debug usage at safepoints.

The downside is that, in general, rematerialization logic has to use the corresponding pure function in order to materialize the eliminated instance. In this particular case (primitive boxing), it has to take into account the caching effects of primitive box factories. Otherwise, user code can encounter identity paradoxes with rematerialized primitive box instances.

I don't see how the scalarization logic you propose preserves identity constraints imposed by `valueOf` factories.

navyxliu commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,53 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+        uint first_ind = (uncommon_trap_node->req() - uncommon_trap_node->jvms()->scloff());
+        Node* sobj = new SafePointScalarObjectNode(gvn.type(res)->isa_oopptr(),
+#ifdef ASSERT
+                                                  (AllocateNode*)call,
```
You can use call->isa_Allocate();  It utilizes node's ad-hoc RTTI  to do type casting.

Wanghuang-Huawei replied:
OK. I have done that.

vnkozlov commented:
> The improvement you are proposing is not specific to uncommon traps, but can be generalized to any debug usage at safepoints.
> 
> The downside is that, in general, rematerialization logic has to use the corresponding pure function in order to materialize the eliminated instance. In this particular case (primitive boxing), it has to take into account the caching effects of primitive box factories. Otherwise, user code can encounter identity paradoxes with rematerialized primitive box instances.
> 
> I don't see how the scalarization logic you propose preserves identity constraints imposed by `valueOf` factories.

Yes, it seems this optimization introduces the issue we had with Graal (8223320):
"C2 doesn't model Integer.valueOf() as anything special. It just inlines it. So the check that determines whether to allocate a new Integer or take one from the cache always happens at runtime. Graal models it as a BoxNode. It is correctly lowered, however, if it needs to be present in a JVM state, it is described as an allocation. So the decision whether to allocate or take the cached value has to happen during the deopt."
There is code in deoptimizer for JVMCI which looks for cached Boxed values. We may need to adopt it for C2 EA for this optimization to work.

veresov commented:
> > The improvement you are proposing is not specific to uncommon traps, but can be generalized to any debug usage at safepoints.
> > The downside is that, in general, rematerialization logic has to use the corresponding pure function in order to materialize the eliminated instance. In this particular case (primitive boxing), it has to take into account the caching effects of primitive box factories. Otherwise, user code can encounter identity paradoxes with rematerialized primitive box instances.
> > I don't see how the scalarization logic you propose preserves identity constraints imposed by `valueOf` factories.
> 
> Yes, it seems this optimization introduces the issue we had with Graal (8223320):
> "C2 doesn't model Integer.valueOf() as anything special. It just inlines it. So the check that determines whether to allocate a new Integer or take one from the cache always happens at runtime. Graal models it as a BoxNode. It is correctly lowered, however, if it needs to be present in a JVM state, it is described as an allocation. So the decision whether to allocate or take the cached value has to happen during the deopt."
> There is code in deoptimizer for JVMCI which looks for cached Boxed values. We may need to adopt it for C2 EA for this optimization to work.

In addition to use the logic that we already have there for Graal, you need to track where the box came from. If it comes as a result of valueOf() then it has to come from the caches, if it's somethings that the user allocates with ```new Integer(0)```, then it should be just a normal scalarized object.

veresov commented:
Also, I don't know if EA can handle a case when an Integer that is coming from an allocation and from a valueOf() are both inputs to a phi.
```
Integer i = p ? new Integer(i1) : Integer.valueOf(i2);
deopt();
int use = i.intValue();
```
If it does, then we'd need to force materialization.

vnkozlov commented:
> Also, I don't know if EA can handle a case when an Integer that is coming from an allocation and from a valueOf() are both inputs to a phi.
> 
> ```
> Integer i = p ? new Integer(i1) : Integer.valueOf(i2);
> deopt();
> int use = i.intValue();
> ```
> 
> If it does, then we'd need to force materialization.

Currently C2 EA can't scalarize merged allocations. So this is not an issue for now but may be in a future. Also C2 does not replace valueOf() with scalar node - it inlines it. As result you have branches with allocation and load from cache.

Only this patch propose to use scalar node for valueOf() for the first time in C2. But it is replaced only in case it is directly referenced  by debug info (no Phi or other nodes in between). So I think it is safe if we use AutoBoxObjectValue for it.

vnkozlov reviewed (commented):
Please, add test case which verifies that Box is scalarized by forking process and checking output of run with `-XX:+PrintEliminateAllocations` flag.
You also need a test which triggers deoptimization and execute code for Box object reallocation/initialization or load from cache.
A test should also verifies that box object identity matches after deoptimization in case when object is loaded from cache.

TobiHartmann commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,56 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+static void replace_box_to_scalar(CallNode* call, Node* resproj) {
+  if (resproj != nullptr && call->is_CallStaticJava() &&
+      call->as_CallStaticJava()->is_boxing_method()) {
+    Unique_Node_List debuginfo_node_list;
```
Maybe rename this to `safepoints`.

Wanghuang-Huawei replied:
Sure.

TobiHartmann commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,56 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+        uint dbg_start = sfpt->is_Call() ? sfpt->as_Call()->tf()->domain()->cnt() : (uint)TypeFunc::Parms+1;
+        for (uint i = 0; i < dbg_start; i++) {
+          if (sfpt->in(i) == resproj) {
+            return;
```
I think this code can be replaced by:
```
  if (!sfpt->is_Call() || !sfpt->as_Call()->has_non_debug_use(n)) {
    safepoints.push(sfpt);
  } else {
    ...
  }
```

Wanghuang-Huawei replied:
Yes, I will do that.

TobiHartmann commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,56 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+
+      ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
+      int n_fields = klass->nof_nonstatic_fields();
+      assert(n_fields == 1, "the klass must be an auto-boxing klass");
```
This code can be put in `ifdef ASSERT` and `n_fields` below can be replaced by 1.

Wanghuang-Huawei replied:
These codes are similar to https://github.com/openjdk/jdk/pull/853#discussion_r522967411 , so I put `assert` here.

TobiHartmann commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -603,6 +653,8 @@ void CallGenerator::do_late_inline_helper() {
     C->remove_macro_node(call);
   }
 
+  replace_box_to_scalar(call, callprojs.resproj);
```
Should this be guarded by `C->eliminate_boxing()`?

Wanghuang-Huawei replied:
Exactly. I will change that.

TobiHartmann commented on src/hotspot/share/opto/callnode.hpp:
```diff
@@ -500,6 +500,7 @@ class SafePointScalarObjectNode: public TypeNode {
                      // states of the scalarized object fields are collected.
                      // It is relative to the last (youngest) jvms->_scloff.
   uint _n_fields;    // Number of non-static fields of the scalarized object.
+  bool _is_auto_box; // is the scalarized object is auto box.
```
Typo in comment. Should be something like `// True if the scalarized object is an auto box`

Wanghuang-Huawei replied:
Thank you for your review. I'll add this in my next patch.

TobiHartmann commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,56 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+    // delay box in runtime, treat box as a scalarized object
+    while (debuginfo_node_list.size() > 0) {
+      ProjNode* res = resproj->as_Proj();
+      Node* debuginfo_node = debuginfo_node_list.pop();
```
`debuginfo_node` -> `safepoint`

Wanghuang-Huawei replied:
OK.

TobiHartmann commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,56 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+#endif // ASSERT
+                                                  first_ind, n_fields, true);
+      sobj->init_req(0, kit.root());
+      debuginfo_node->add_req(call->in(res->_con));
```
I don't understand why you are selecting the input based on the result projection field `res->_con`?

Wanghuang-Huawei replied:
Thank you for your review. I use `sfpt->add_req(call->in(TypeFunc::Parms));` instead of this.

TobiHartmann reviewed (changes requested):
I agree with Vladimir that tests are needed. I've added some comments to the code.

Wanghuang-Huawei reviewed (commented):
> Please, add test case which verifies that Box is scalarized by forking process and checking output of run with `-XX:+PrintEliminateAllocations` flag.
> You also need a test which triggers deoptimization and execute code for Box object reallocation/initialization or load from cache.
> A test should also verifies that box object identity matches after deoptimization in case when object is loaded from cache.

OK. I will add these test cases.

vnkozlov reviewed (approved):
This looks good. I will test it and let you know results.

vnkozlov commented:
Got failure during testing `applications/ctw/modules/java_desktop.java` on linux-aarch64.
With additional flags `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation -Djava.awt.headless=true`

```
#  Internal Error (/workspace/open/src/hotspot/share/opto/loopTransform.cpp:3711), pid=305808, tid=305821
#  assert(use == polladr) failed: the use should be a safepoint polling
#

Current CompileTask:
C2:  66678 5846   !b        com.sun.media.sound.RealTimeSequencer$DataPump::makeDisabledArray (183 bytes)

Stack: [0x0000ffff53c00000,0x0000ffff53e00000],  sp=0x0000ffff53df9c50,  free space=2023k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x11d7090]  PhaseIdealLoop::match_fill_loop(IdealLoopTree*, Node*&, Node*&, Node*&, Node*&)+0xbe0
V  [libjvm.so+0x11d7388]  PhaseIdealLoop::intrinsify_fill(IdealLoopTree*) [clone .part.0]+0xa4
V  [libjvm.so+0x11d81e8]  PhaseIdealLoop::do_intrinsify_fill()+0x218
V  [libjvm.so+0x120afb8]  PhaseIdealLoop::build_and_optimize(LoopOptsMode)+0xec8
V  [libjvm.so+0x98027c]  PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x28c
V  [libjvm.so+0x97d9bc]  Compile::Optimize()+0xf3c
V  [libjvm.so+0x97ed20]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0xf50
```

Wanghuang-Huawei commented:
> Got failure during testing `applications/ctw/modules/java_desktop.java` on linux-aarch64.
> With additional flags `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation -Djava.awt.headless=true`
> 
> ```
> #  Internal Error (/workspace/open/src/hotspot/share/opto/loopTransform.cpp:3711), pid=305808, tid=305821
> #  assert(use == polladr) failed: the use should be a safepoint polling
> #
> 
> Current CompileTask:
> C2:  66678 5846   !b        com.sun.media.sound.RealTimeSequencer$DataPump::makeDisabledArray (183 bytes)
> 
> Stack: [0x0000ffff53c00000,0x0000ffff53e00000],  sp=0x0000ffff53df9c50,  free space=2023k
> Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
> V  [libjvm.so+0x11d7090]  PhaseIdealLoop::match_fill_loop(IdealLoopTree*, Node*&, Node*&, Node*&, Node*&)+0xbe0
> V  [libjvm.so+0x11d7388]  PhaseIdealLoop::intrinsify_fill(IdealLoopTree*) [clone .part.0]+0xa4
> V  [libjvm.so+0x11d81e8]  PhaseIdealLoop::do_intrinsify_fill()+0x218
> V  [libjvm.so+0x120afb8]  PhaseIdealLoop::build_and_optimize(LoopOptsMode)+0xec8
> V  [libjvm.so+0x98027c]  PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x28c
> V  [libjvm.so+0x97d9bc]  Compile::Optimize()+0xf3c
> V  [libjvm.so+0x97ed20]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0xf50
> ```


OK . I will check this error.

Wanghuang-Huawei commented:
> Got failure during testing `applications/ctw/modules/java_desktop.java` on linux-aarch64.
> With additional flags `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation -Djava.awt.headless=true`
> 
> ```
> #  Internal Error (/workspace/open/src/hotspot/share/opto/loopTransform.cpp:3711), pid=305808, tid=305821
> #  assert(use == polladr) failed: the use should be a safepoint polling
> #
> 
> Current CompileTask:
> C2:  66678 5846   !b        com.sun.media.sound.RealTimeSequencer$DataPump::makeDisabledArray (183 bytes)
> 
> Stack: [0x0000ffff53c00000,0x0000ffff53e00000],  sp=0x0000ffff53df9c50,  free space=2023k
> Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
> V  [libjvm.so+0x11d7090]  PhaseIdealLoop::match_fill_loop(IdealLoopTree*, Node*&, Node*&, Node*&, Node*&)+0xbe0
> V  [libjvm.so+0x11d7388]  PhaseIdealLoop::intrinsify_fill(IdealLoopTree*) [clone .part.0]+0xa4
> V  [libjvm.so+0x11d81e8]  PhaseIdealLoop::do_intrinsify_fill()+0x218
> V  [libjvm.so+0x120afb8]  PhaseIdealLoop::build_and_optimize(LoopOptsMode)+0xec8
> V  [libjvm.so+0x98027c]  PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x28c
> V  [libjvm.so+0x97d9bc]  Compile::Optimize()+0xf3c
> V  [libjvm.so+0x97ed20]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0xf50
> ```

Does this error reproduct every time? I tested this case in my aarch64 machine for many times （> 300） and failed to encounter this error. 

Can you give me some more info about this error? For example, crash log ? 

PS: 
I used `make run-test CONF=slowdebug JTREG="JAVA_OPTIONS=-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation -Djava.awt.headless=true" TEST="applications/ctw/modules/java_desktop.java"` to run this case.

vnkozlov commented:
I attached hs_err and replay files to [8261137](https://bugs.openjdk.java.net/browse/JDK-8261137).
It happens only once. Only on aarch64.

Wanghuang-Huawei commented:
> I attached hs_err and replay files to [8261137](https://bugs.openjdk.java.net/browse/JDK-8261137).
> It happens only once. Only on aarch64.

* Good News :  

    * I have reproducted this error on my machine. 
    *  I reverted my commit and could also reproduct this error.  It seems that this error has nothing to do with my commit.

* Bad News:
  *   I have not found the reason. 
  *   It seems that this error is a random bug. After patching my commit on the newest jdk/jdk, I could reproduct this error for one time in 40 tests.

vnkozlov commented:
1:40 ratio is good :) Did you try replay file?
Please, file new bug and describe how you reproduced it (without patch).

About these changes. I am fine with them but you need second review for final code.

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -603,6 +655,10 @@ void CallGenerator::do_late_inline_helper() {
   }
 
+  if (C->eliminate_boxing()) {
+    replace_box_to_scalar(call, callprojs.resproj);
```
Another confusing aspect: what happens with `call` node after `replace_box_to_scalar` call is over?

My understanding is you rely on `replace_box_to_scalar` to eliminate all the uses by safepoints and then let  the following code to clean it up:
```
  bool result_not_used = (callprojs.resproj == NULL || callprojs.resproj->outcnt() == 0);
  if (is_pure_call() && result_not_used) {
    // The call is marked as pure (no important side effects), but result isn't used.
    // It's safe to remove the call.
    GraphKit kit(call->jvms());
    kit.replace_call(call, C->top(), true);
```

The relation between `is_pure_call()` and `C->eliminate_boxing()`/`call->as_CallStaticJava()->is_boxing_method()` is not clear.

I suggest to refactor `is_pure_call()` and add a special case for `is_boxing_method()` to perform scalarization when there are only debug usages. 

Also, add an assert in/after `replace_box_to_scalar` ensuring there are no users left, so the call can go away.

iwanowww replied:
Regarding `C->eliminate_boxing()`: why do you think it is needed? 

I'm asking because `is_pure_call() && result_not_used` doesn't have that check.

My understanding is the code encounters `is_boxing_method()` there only when `Compile::should_delay_boxing_inlining() == true` which requires both `EliminateAutoBox` and `AggressiveUnboxing`

Wanghuang-Huawei replied:
I will refactor these codes and add `is_boxing_call()` method. Thank you very much for your review.

vnkozlov replied:
I just hit bug where VectorReinterpretTest.java test run with -XX:-DoEscapeAnalysis but Vector API generates SafePointScalarObjectNode regardless that:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vector.cpp#L249

But during deoptimization we only restore allocations when DoEscapeAnalysis is true:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/deoptimization.cpp#L377

(currently condition is always true because jvmci_enabled == true since JVMCI included into build but without JVMCI it will fail as I found).

It seems with this change you need to add EliminateAutoBox flag check to deoptimization condition.
Or guard your optimization with C->do_escape_analysis().

@iwanowww  Do you think we should add checks for EnableVectorAggressiveReboxing and  EliminateAutoBox into deoptimization code or guard Vector API code and this optimization with  C->do_escape_analysis()?

vnkozlov replied:
> I just hit bug where VectorReinterpretTest.java test run with -XX:-DoEscapeAnalysis but Vector API generates SafePointScalarObjectNode regardless that:
> https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vector.cpp#L249
> 
> But during deoptimization we only restore allocations when DoEscapeAnalysis is true:
> https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/deoptimization.cpp#L377
> 
> (currently condition is always true because jvmci_enabled == true since JVMCI included into build but without JVMCI it will fail as I found).
> 
> It seems with this change you need to add EliminateAutoBox flag check to deoptimization condition.
> Or guard your optimization with C->do_escape_analysis().
> 
> @iwanowww Do you think we should add checks for EnableVectorAggressiveReboxing and EliminateAutoBox into deoptimization code or guard Vector API code and this optimization with C->do_escape_analysis()?

The issue was fixed: 8263125. https://git.openjdk.java.net/jdk/pull/2924

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,58 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+    }
+
+#ifndef PRODUCT
+    if (PrintEliminateAllocations && safepoints.size() > 0) {
```
Strictly speaking, `PrintEliminateAllocations` is EA-specific flag.
Not sure it has to cover `EliminateAutoBox`/`AggressiveUnboxing`.

Wanghuang-Huawei replied:
I think it is just like  https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/macro.cpp#L1158
or we add a new option here?

iwanowww replied:
I missed that `PrintEliminateAllocations` is used in `PhaseMacroExpand::eliminate_boxing_node()`. 

Looks good then.

Wanghuang-Huawei commented:
> I attached hs_err and replay files to [8261137](https://bugs.openjdk.java.net/browse/JDK-8261137).
> It happens only once. Only on aarch64.

I have found the reason of this failure. I will fix it under this issue: https://bugs.openjdk.java.net/browse/JDK-8263352
* JDK-8260637 make a `CastPP` unpinned, so it can be pull out of inner loop
* JDK-8247307 assert `use == polladr` and makes it crash here.
* There are some details which we can discuss under the issue JDK-8263352. For example , do we need this assert,  OR,  should we fix JDK-8260637 in other way?
Luckly, it has no issue with my patch. :-)

Wanghuang-Huawei commented on test/hotspot/jtreg/compiler/c2/TestIdentityWithEliminateBoxInDebugInfo.java:
```diff
@@ -0,0 +1,113 @@
+ * @summary Verify that box object identity matches after deoptimization When it is eliminated.
+ * @library /test/lib
+ *
+ * @run main/othervm compiler.c2.TestIdentityWithEliminateBoxInDebugInfo
```
Sure. Thank you.

vnkozlov commented:
> > I attached hs_err and replay files to [8261137](https://bugs.openjdk.java.net/browse/JDK-8261137).
> > It happens only once. Only on aarch64.
> 
> I have found the reason for this failure. I will fix it under this issue: https://bugs.openjdk.java.net/browse/JDK-8263352
> 
> * JDK-8260637 make a `CastPP` unpinned, so it can be pull out of inner loop
> * JDK-8247307 assert `use == polladr` and makes it crash here.
> * There are some details which we can discuss under the issue JDK-8263352. For example , do we need this assert,  OR,  should we fix JDK-8260637 in other way?
>   Luckly, it has no issue with my patch. :-)

Yes, it is good news. Thank you  tracking it down.

vnkozlov commented:
We would also need to check EliminateAllocations flag creating generating SafePointScalarObjectNode :(
May it is easy to update checks in deoptimizer.

vnkozlov commented:
> > @iwanowww Do you think we should add checks for EnableVectorAggressiveReboxing and EliminateAutoBox into deoptimization code or guard Vector API code and this optimization with C->do_escape_analysis()?
> 
> The former (additional checks for EnableVectorAggressiveReboxing and EliminateAutoBox).
> Vector box elimination is unrelated to escape analysis pass and should work irrespective of whether EA is turned on or off.

Okay. I will do it for 8263125 fix.

vnkozlov commented:
I added both flags checks in deoptimizer in https://github.com/openjdk/jdk/pull/2924
@Wanghuang-Huawei you don't need to do it here.

But I am lost on the status of these changes. Do you have other updates for them? Or current 05 is final?
You found that the testing failure is not related and addressed other my and @iwanowww questions.
If 05 version is final and I already tested it, we are ready for approval.

Wanghuang-Huawei commented:
> I added both flags checks in deoptimizer in #2924
> @Wanghuang-Huawei you don't need to do it here.
> 
> But I am lost on the status of these changes. Do you have other updates for them? Or current 05 is final?
> You found that the testing failure is not related and addressed other my and @iwanowww questions.
> If 05 version is final and I already tested it, we are ready for approval.

OK. Thank you for your work.
I will update some codes to answer the review  of  @iwanowww. It will be 06 version.

Wanghuang-Huawei commented:
@iwanowww Can you review my new 06 version? I have split `is_boxing_call` from `is_pure_call`. :-)

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -728,7 +800,7 @@ class LateInlineBoxingCallGenerator : public LateInlineCallGenerator {
  public:
   LateInlineBoxingCallGenerator(ciMethod* method, CallGenerator* inline_cg) :
-    LateInlineCallGenerator(method, inline_cg, /*is_pure=*/true) {}
+    LateInlineCallGenerator(method, inline_cg, /*is_pure=*/true, /*is_boxing*/true) {}
```
You can replace `_is_boxing_call` with `is_boxing_late_inline()`. I find it clearer.

```
src/hotspot/share/opto/callGenerator.cpp
@@ -813,6 +795,8 @@ class LateInlineBoxingCallGenerator : public LateInlineCallGenerator {
     return new_jvms;
   }
 
+  virtual bool is_boxing_late_inline() const { return true; }
+
   virtual CallGenerator* with_call_node(CallNode* call) {
     LateInlineBoxingCallGenerator* cg = new LateInlineBoxingCallGenerator(method(), _inline_cg);
     cg->set_call_node(call->as_CallStaticJava());

src/hotspot/share/opto/callGenerator.hpp
@@ -75,6 +75,7 @@ class CallGenerator : public ResourceObj {
   virtual bool      is_late_inline() const         { return false; }
   // same but for method handle calls
   virtual bool      is_mh_late_inline() const      { return false; }
+  virtual bool      is_boxing_late_inline() const  { return false; }
   virtual bool      is_string_late_inline() const  { return false; }
   virtual bool      is_virtual_late_inline() const { return false; }
```

Wanghuang-Huawei replied:
Thank you for your review. I fix it in my patch.

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -603,10 +666,19 @@ void CallGenerator::do_late_inline_helper() {
+
+  if (is_pure_call()) {
+    if (is_boxing_call() && callprojs.resproj != nullptr) {
+        replace_box_to_scalar(call, callprojs.resproj);
```
I suggest to split it into 2 parts: 
  * `bool has_non_debug_usages(Node* n)` check which ensures there are only debug usages present
  *  `scalarize_debug_usages(...)` which iterates over all usages assuming they are debug and scalarizes them.

After that you can just keep `result_not_used` as is. After all debug usages are gone, the call becomes clearly not used. And IMO `result_not_used` communicates the intention better than `call_can_eliminate`.

Wanghuang-Huawei replied:
Sure, I refact these codes.

iwanowww reviewed (commented):
Overall, looks good.

Some refactoring suggestions.

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,62 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+
+static void scalarize_debug_usages(CallNode* call, Node* resproj) {
+  Unique_Node_List safepoints;
+  for (DUIterator_Fast imax, i = resproj->fast_outs(imax); i < imax; i++) {
```
No need to collect safepoints anymore. You can process users while iterating over them.
```
for (DUIterator_Last imin, i = res->last_outs(imin); i >= imin;) {
  SafePointNode* sfpt = res->last_out(i)->as_SafePoint();
  ...
  int num_edges = sfpt->replace_edges_in_range(res, sobj, start, end);  
  i -= num_edges;
}
```

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,62 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+  }
+
+#ifndef PRODUCT
+  if (PrintEliminateAllocations && safepoints.size() > 0) {
```
`safepoints.size() > 0` is redundand.

Also, I'd move it to the end of the method after all the users are processed.

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,62 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+    ProjNode* res = resproj->as_Proj();
+    Node* sfpt = safepoints.pop();
+
+    ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
```
`klass` is loop invariant and can be moved out of the loop (along with `n_fields`).

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -554,6 +554,62 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
+    uint first_ind = sfpt->req() - sfpt->jvms()->scloff();
+    Node* sobj = new SafePointScalarObjectNode(gvn.type(res)->isa_oopptr(),
+#ifdef ASSERT
+                                                  call->isa_Allocate(),
```
It is always `NULL` since `call` can't be an `Allocate`.

vnkozlov replied:
It was my suggestion to pass `call` node as allocation so that we could trace back for what node SafePointScalarObject was created because you may have several Box objects for which we create SafePointScalarObject nodes.
I think we can change argument (and field type) to CallNode. And have assert in SafePointScalarObject constructor to check that it is either Allocate node or Boxing Call node.

Wanghuang-Huawei replied:
Thank you for your review. I will change the  SafePointScalarObject constructor in next push.

iwanowww reviewed (approved):
Looks good.

Some more minor cleanup suggestions follow.

Wanghuang-Huawei commented on test/hotspot/jtreg/compiler/c2/TestEliminateBoxInDebugInfo.java:
```diff
@@ -0,0 +1,88 @@
+/*
+ * @test
+ * @bug 8261137
+ * @requires vm.debug == true & vm.flavor == "server"
```
OK. I will do that ASAP.

Wanghuang-Huawei commented on test/hotspot/jtreg/compiler/c2/TestEliminateBoxInDebugInfo.java:
```diff
@@ -0,0 +1,88 @@
+/*
```
OK. I will do that.

Wanghuang-Huawei commented on test/hotspot/jtreg/compiler/c2/TestIdentityWithEliminateBoxInDebugInfo.java:
```diff
@@ -0,0 +1,113 @@
+/*
```
OK. Thank you for your review.

vnkozlov commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -1515,17 +1515,24 @@ void SafePointNode::disconnect_from_root(PhaseIterGVN *igvn) {
 {
+#ifdef ASSERT
+  if (alloc != nullptr) {
+    assert(alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected alloc");
```
Add dumping of `alloc` **before** this assert to know it during failure without rerunning test:
```
if (!alloc->is_Allocate() && !alloc->as_CallStaticJava()->is_boxing_method()) {
  alloc->dump();
}
```

iwanowww replied:
> It was my suggestion to pass call node as allocation so that we could trace back for what node SafePointScalarObject was created because you may have several Box objects for which we create SafePointScalarObject nodes.

The problem with that is `alloc` quickly becomes stale since allocations/calls are removed shortly after scalarization takes place. The only usage of `alloc()` is in `PhaseMacroExpand::scalar_replacement`:
```
assert(scobj->alloc() == alloc, "sanity");
```

Regarding the assert, frankly speaking, it looks repetitive and verbose. I'd prefer to just see it folded into:
```
assert(alloc == NULL || alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected alloc");
```

Or introduce a helper method to dump relevant info about the problematic node: 
```
assert(alloc == NULL || alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected alloc: %s", dump_node(alloc));
```

Also, `alloc == NULL` case can be eliminated by avoiding passing `NULL` in `PhaseVector::scalarize_vbox_node()`.

vnkozlov replied:
Okay. Lets do as @iwanowww suggesting.

Wanghuang-Huawei replied:
> Okay. Lets do as @iwanowww suggesting.

Should we implement `dump_node` or should we just use `assert(alloc == NULL || alloc->is_Allocate() || alloc->as_CallStaticJava()->is_boxing_method(), "unexpected call node");`  without dumping?

vnkozlov replied:
Creating node's dump string in separate method is complicated.
We can just use `false` in assert:
```
#ifdef ASSERT
  if (alloc != nullptr && !alloc->is_Allocate() && !alloc->as_CallStaticJava()->is_boxing_method()) {
    alloc->dump();
    assert(false, "unexpected call node for scalar replacement");
  }
#endif
```

vnkozlov replied:
I did not see your latest changes before answering your question. You did exactly as I suggested. Good.

Wanghuang-Huawei replied:
Thank you for your review.

vnkozlov reviewed (approved):
Good.

Wanghuang-Huawei commented:
> Good.

Thank you for your approval. ;-)

Wanghuang-Huawei commented:
/integrate

Wanghuang-Huawei commented:
@vnkozlov Can you do me a favor to `/sponsor` this patch?

TobiHartmann reviewed (approved):
Looks good to me too.

vnkozlov commented:
I will wait results of testing Vladimir I. is running.

iwanowww commented:
Test results (hs-tier1 - hs-tier4) are clean.

/sponsor
